src/macros-by-example.adoc

```diff
@@ -0,0 +1,826 @@
+----
+
+The delimiting sequence is effectively shorthand for a `values` expression, but arbitrary
+expressions are not allowed:
+
+[{mrk}]
+----
+(:prices (:values 10 9.99 12.) GBP) ⇒ _**error**: delimiting list or sexp expected_
+(:prices (:) GBP)                   ⇒ _**error**: delimiting list or sexp expected_
+----
+
+That’s because the binary representation of these parameters uses a tagless format for these
+delimiters to keep the common case as dense as possible, and it’s not possible to replace that
+container with a macro invocation.  The text form mirrors that limitation, but you can always
+call a macro inside the delimiter, with no loss of generality:
+
+----
+(:prices [(:values 10)] GBP) ⇒ {amount:10, currency:GBP}
+----
+
+Note that the delimiter is required even for single values, again for consistency but more
+importantly to avoid ambiguity.  Consider this macro:
+
+[{mrk}]
+----
+(*public* ouch [(*list** stuff)] …)
+----
+
+In the E-expression `(:ouch [])`  is that list the delimiting container, or the sole input value?
+When expanding the E-expression does `stuff` produce void, or a single empty list?  `***`-arguments
+require a delimiting container precisely so there’s no ambiguity, regardless of the element type,
+and the answer is that `stuff` is void.
+
+Within the delimiting container, the invocation can have any number of values and/or macro
+invocations.  The macro parameter produces the results of those expressions, concatenated into a
+single stream, and the expander verifies that each value on that stream is acceptable by the
+parameter’s declared type.
+
+----
+(:prices (1 (:values 2 3) 4) GBP) ⇒ {amount:1, currency:GBP}
+                                    {amount:2, currency:GBP}
+                                    {amount:3, currency:GBP}
+                                    {amount:4, currency:GBP}
+----
+
+
+==== one-or-more
+
+A one-or-more `*+*` parameter works just like `***` except the resulting stream must produce at
+least one value.  To continue abusing our prices example:
+
+[{mrk}]
+----
+(*public* prices
+  [(*number+* amount), (*symbol* currency)]
+  (*for_each* amount
+    (price amount currency)))
+----
+
+[{mrk}]
+----
+(:prices () JPY) ⇒ _**error**: at least one value expected for + parameter_
+(:prices [99] EUR)               ⇒ {amount:99, currency:EUR}
+(:prices \((:) (:values 99)) EUR) ⇒ {amount:99, currency:EUR}
+----
+
+Note that among the delimited expressions, empty results are fine, as long as their concatenation
+produces one or more values.
+
+
+=== Optional Arguments
+
+When a trailing parameter is voidable, an invocation can omit its corresponding argument, as long
+as no following parameter is being given a value.  We’ve seen this as applied to rest-parameters,
+but it also applies to `*?*` and `***` cardinality:
+
+[{mrk}]
+----
+(*public* optionals
+  [(*any** a), (*any?* b), (*any!* c), (*any** d), (*any?* e), (*any\...* f)]
```

Comment:
The `*` escaping is getting screwed up on this line. I'm seeing `*any?` in the rendered view, and most of it is bold.

Reply:
Fixed.